yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+            logDebug(s"About to POST $entityDescription")
+            try { 
+              val response: TimelinePutResponse = client.putEntities(entity)
+              val errors = response.getErrors
+              if (!response.getErrors.isEmpty) {
+                val err = errors.get(0)
+                eventPostFailures.addAndGet(errors.size())
+                if (err.getErrorCode != 0) {
+                  logError(s"Failed to post ${entityDescription}\n:${describeError(err)}")
+                }
+              } else {
+                // successful submission
+                logDebug(s"entity successfully posted: $entityDescription")
+              }
+              // whatever the outcome, this request is not re-issued
+              false
+            } catch {
+              case e: ConnectException =>
+                eventPostFailures.incrementAndGet
+                logWarning(s"Connection exception submitting $entityDescription\n$e", e)
+                true
+
+              case e: RuntimeException =>
+                // this is probably a retry timeout event; some Hadoop versions don't
+                // rethrow the exception causing the problem, instead raising an RTE
+                eventPostFailures.incrementAndGet
+                logWarning(s"Runtime exception submitting $entityDescription\n$e", e)
+                // same policy as before: retry on these
+                true
+
+              case e: Exception =>
+                // something else has gone wrong.
+                eventPostFailures.incrementAndGet
+                logWarning(s"Could not handle history entity: $entityDescription\n$e", e)
+                false
+            }
+          }
+        }
+      }
+      logDebug(s"after pushEntities: ${entityList.size}")
+    }
+  }
+
+  /**
+   * Process the action placed in the queue
+   * @param action action
+   * @return true if the queue processor must now exit
+   */
+  private def processAction(action: QueuedAction): Boolean = {
+    action match {
+
+      case StopQueue() =>
+        logDebug("Stop queue action received")
+        true
+
+      case event: HandleSparkEvent =>
+        try {
+          handleEvent(event)
+          false
+        } catch {
+          case NonFatal(e) => false
+        }
+
+      case FlushTimelineEvents() =>
+        logDebug("Flush queue action received")
+        flushCurrentEventsToATS
+        false
+    }
+  }
+
+  /**
+   * If the event reaches the batch size or flush is true, push events to ATS.
+   *
+   * @param event event. If null no event is queued, but the post-queue flush logic
+   *              still applies
+   * @return true if the event should stop the service.
+   */
+  private def handleEvent(event: HandleSparkEvent): Unit = {
+    var push = false
+    // if we receive a new appStart event, we always push
+    // not much contention here, only happens when service is stopped
+    entityLock synchronized {
```

Comment:
it probably dates from a time when the service stop code could also issue operations -culling